RPC: Add reserve member function to `UniValue` and use it in `blockToJSON` function

ismaelsadeeq opened the pull request:
This PR is motivated by https://github.com/bitcoin/bitcoin/issues/30495#issuecomment-2444881418,
It adds a `reserve` member function to `UniValue` and applies it within the `blockToJSON` function to pre-allocate memory, minimizing reallocation's. 

<details>
<summary>Slight performance increase in verbosity 1</summary>

On **master**:

|         ns/op |        op/s |  err% | total | benchmark              |
|--------------:|------------:|------:|------:|:------------------------|
|      190,342  |     5,254   |  2.3% |  0.01 | `BlockToJsonVerbose1`   |
|  34,812,292   |       28.73 |  1.0% |  0.38 | `BlockToJsonVerbose2`   |
|  34,457,167   |       29.02 |  1.0% |  0.38 | `BlockToJsonVerbose3`   |

On **this PR**:

|         ns/op |        op/s |  err% | total | benchmark              |
|--------------:|------------:|------:|------:|:------------------------|
|      172,278  |     5,805   |  0.7% |  0.01 | `BlockToJsonVerbose1`   |
|  33,720,584   |       29.66 |  0.4% |  0.37 | `BlockToJsonVerbose2`   |
|  33,884,417   |       29.51 |  1.2% |  0.38 | `BlockToJsonVerbose3`   |
</details>

ismaelsadeeq committed:
UniValue: add reserve member function

- Only reserve keys when the typ is of `VOBJ`.

ismaelsadeeq committed:
rpc: reserve space for `UniValue` variables  in `blockToJSON`

- Reserving space avoid reallocation, this provide noticeable
  performance increase in verbosity 1.

andrewtoth commented:
Concept ACK

andrewtoth commented:
I think we can take this further and reserve for VOBJ types instead of just VARR.
The VOBJ uses both values and keys, so we can reserve both. We can count how many times we do `pushKV` and reserve that amount for both keys and values.

andrewtoth commented:
I believe you can get more fine grained benchmarks for this using the rpc_blockchain benchmarks.

ismaelsadeeq commented:
> I think we can take this further and reserve for `VOBJ` types instead of just `VARR`. The `VOBJ` type uses both values and keys, so we can reserve for both.

Done in the latest push, see [diff](https://github.com/bitcoin/bitcoin/compare/fa73b9a9984d5dd2efa6fd5dd70b3db617803a25..f8d5d2cfc63bea5b58b08ceaa743c26820c1763d), but I am reserving keys only when `typ` is of `VOBJ`.

> We can count how many times we use `pushKV` and reserve that amount for both keys and values.

+1, this approach is possible but is it maintainable?
Manually counting each `pushKV`s means updating the value thats being passed to `.reserve` every time we add another element, will be better to reserve for size's that are determined at runtime or at least use a constant.

Is there a constant for the number of items in a block so that I can reserve for results in `blockheaderToJSON`?

ismaelsadeeq commented:
> Also, the witness stack array inside TxToUniv is a VARR that could be reserved with a parameter resolved at runtime.

Thanks, I've added this in the latest push.

Eunovo commented:
> > I believe you can get more fine-grained benchmarks for this using the `rpc_blockchain` benchmarks.
> 
> Yes, I can, but I’ll need to add verbosity levels 1 and 2, where we can see some performance improvements.
> 
> Show diff
> Benchmark Results
> 
> On **master**:
> 
> ns/op	op/s	err%	total	benchmark
> 190,342	5,254	2.3%	0.01	`BlockToJsonVerbose1`
> 34,812,292	28.73	1.0%	0.38	`BlockToJsonVerbose2`
> 34,457,167	29.02	1.0%	0.38	`BlockToJsonVerbose3`
> On **this PR**:
> 
> ns/op	op/s	err%	total	benchmark
> 172,278	5,805	0.7%	0.01	`BlockToJsonVerbose1`
> 33,720,584	29.66	0.4%	0.37	`BlockToJsonVerbose2`
> 33,884,417	29.51	1.2%	0.38	`BlockToJsonVerbose3`

@ismaelsadeeq, could you commit the diff you’ve provided and include the benchmark results in the commit message? Then, rebase your current changes on top, rerun the benchmarks for https://github.com/bitcoin/bitcoin/pull/31179/commits/952463ccdc2baeacb527c66732fd184cbcbd53ba and https://github.com/bitcoin/bitcoin/pull/31179/commits/28e3392d11355b1160dc1a7a5557081728a02840, and update their commit messages with the results.

This way I can just go through your commits, run the benchmarks and compare with your results.

ismaelsadeeq committed:
bench: support benching all verbosity of `BlockToJson`

l0rinc commented:
@ismaelsadeeq, I ran your benchmarks from https://github.com/bitcoin/bitcoin/pull/31179#issuecomment-2446067820, but I only see a 7% speedup in the `TxVerbosity::SHOW_TXID` bench - while the rest were exactly the same as before.
But if it's *that* important here, we could easily get a 30% speedup as well there by deduplicating the `GetSerializeSize` calls:
```C++
static int64_t GetBlockWeight(const size_t size_no_witness, const size_t size_with_witness)
{
    return size_no_witness * (WITNESS_SCALE_FACTOR - 1) + size_with_witness;
}
static int64_t GetBlockWeight(const CBlock& block)
{
    const size_t size_no_witness = ::GetSerializeSize(TX_NO_WITNESS(block));
    const size_t size_with_witness = ::GetSerializeSize(TX_WITH_WITNESS(block));
    return GetBlockWeight(size_no_witness, size_with_witness);
}
```
...
```C++
UniValue blockToJSON(BlockManager& blockman, const CBlock& block, const CBlockIndex& tip, const CBlockIndex& blockindex, TxVerbosity verbosity)
{
    UniValue result = blockheaderToJSON(tip, blockindex);

    const size_t size_no_witness = GetSerializeSize(TX_NO_WITNESS(block));
    const size_t size_with_witness = GetSerializeSize(TX_WITH_WITNESS(block));
    result.pushKV("strippedsize", (int)size_no_witness);
    result.pushKV("size", (int)size_with_witness);
    result.pushKV("weight", (int)GetBlockWeight(size_no_witness, size_with_witness));
    UniValue txs(UniValue::VARR);
    txs.reserve(block.vtx.size());
...
```

----

Or if you need `TxVerbosity::SHOW_DETAILS` and `TxVerbosity::SHOW_DETAILS_AND_PREVOUT`, they can be sped up by ~20% by reviving the abandoned (for lack of interest) https://github.com/bitcoin/bitcoin/pull/30035.

---

And if `BlockToJsonVerboseWrite` needs a boost, https://github.com/bitcoin/bitcoin/pull/31144 will speed up by ~14%.

l0rinc commented:
@ismaelsadeeq https://github.com/bitcoin/bitcoin/pull/20999 by @martinus also contains a few optimization attempts (a lot of them were already applied in other PRs since, but a few of the changes can still be done).
If RPC speed is important to you, the above suggestions can result in a measurable speedups.

ismaelsadeeq commented:
Thanks for taking a look @Eunovo @l0rinc 

> could you commit the diff you’ve provided and include the benchmark results in the commit message?

I've added 8d4823b383be4f8cd0152176a4f67b447d29549 For the benchmarks and instead update the PR title with the bench results.

> I ran your benchmarks from https://github.com/bitcoin/bitcoin/pull/31179#issuecomment-2446067820, but I only see a 7% speedup in the TxVerbosity::SHOW_TXID bench - while the rest were exactly the same as before.

I did that locally and the result is the same as well

>  But if it's that important, we could easily get a 30% speedup as well there by deduplicating the GetSerializeSize calls:

I'll be happy to review that as a PR or might attempt it myself.

> Or if you need TxVerbosity::SHOW_DETAILS and TxVerbosity::SHOW_DETAILS_AND_PREVOUT, they can be sped up by ~20% by reviving the abandoned (for lack of interest) https://github.com/bitcoin/bitcoin/pull/30035.

Interesting I will take a look at that, it seems there are some unaddressed comments by @TheCharlatan at that PR

> And if BlockToJsonVerboseWrite needs a boost, https://github.com/bitcoin/bitcoin/pull/31144 will speed up by ~14%.

I will take a look.

> @martinus also contains a few optimization attempts (a lot of them were already applied in other PRs since, but a few of the changes can still be done).
If RPC speed is important to you, the above suggestions can result in a measurable improvements.

RPC speed is important not just to me but to all users. For now, I've been experimenting with using libbitcoinkernel, and it seems to be more performant than the RPC.

I'll look into it further.

All those PRs offer improvements that are beneficial to users. I also think having this reservation ability is important to avoid the reallocation overhead. @l0rinc?

l0rinc commented:
I can implement my suggestions in separate PRs if you say the RPC speedups would be welcome

l0rinc reviewed (commented):
Approach ACK
There are a LOT of other places where we will be able to use this - we should do them in follow-up PRs

l0rinc commented on src/bench/rpc_blockchain.cpp:
```diff
@@ -45,16 +45,33 @@ struct TestBlockAndIndex {
 }
 
-BENCHMARK(BlockToJsonVerbose, benchmark::PriorityLevel::HIGH);
+static void BlockToJsonVerbose1(benchmark::Bench& bench)
```
the real RPC also [reads the block from disk](https://github.com/bitcoin/bitcoin/blob/master/src/rest.cpp#L322), we could bench that as well - instead of assuming it's already in memory (this will also enable additional optimization opportunities)

ismaelsadeeq replied:
I think this is benching just `blockToJSON` .
But it's a good idea to write a bench for `ReadRawBlockFromDisk` as well.

I will leave that as an idea for a follow-up PR.

l0rinc replied:
> I will leave that as an idea for a follow-up PR.

The problem is that the goal of this PR was to speed up the RPC.
But if we benchmark that instead of just a subset of the task, i.e.:
```patch
diff --git a/src/bench/rpc_blockchain.cpp b/src/bench/rpc_blockchain.cpp
--- a/src/bench/rpc_blockchain.cpp	(revision bf5c569898d0297de010102a623bf52009607ed8)
+++ b/src/bench/rpc_blockchain.cpp	(date 1735836623994)
@@ -24,7 +24,6 @@
 namespace {
 
 struct TestBlockAndIndex {
-    const std::unique_ptr<const TestingSetup> testing_setup{MakeNoLogFileContext<const TestingSetup>(ChainType::MAIN)};
     CBlock block{};
     uint256 blockHash{};
     CBlockIndex blockindex{};
@@ -47,9 +46,10 @@
 
 static void BlockToJson(benchmark::Bench& bench, TxVerbosity verbosity)
 {
-    TestBlockAndIndex data;
+    const std::unique_ptr<const TestingSetup> testing_setup{MakeNoLogFileContext<const TestingSetup>(ChainType::MAIN)};
     bench.run([&] {
-        auto univalue = blockToJSON(data.testing_setup->m_node.chainman->m_blockman, data.block, data.blockindex, data.blockindex, verbosity);
+        TestBlockAndIndex data{};
+        auto univalue = blockToJSON(testing_setup->m_node.chainman->m_blockman, data.block, data.blockindex, data.blockindex, verbosity);
         ankerl::nanobench::doNotOptimizeAway(univalue);
     });
 }
@@ -75,8 +75,9 @@
 
 static void BlockToJsonVerboseWrite(benchmark::Bench& bench)
 {
+    const std::unique_ptr<const TestingSetup> testing_setup{MakeNoLogFileContext<const TestingSetup>(ChainType::MAIN)};
     TestBlockAndIndex data;
-    auto univalue = blockToJSON(data.testing_setup->m_node.chainman->m_blockman, data.block, data.blockindex, data.blockindex, TxVerbosity::SHOW_DETAILS_AND_PREVOUT);
+    auto univalue = blockToJSON(testing_setup->m_node.chainman->m_blockman, data.block, data.blockindex, data.blockindex, TxVerbosity::SHOW_DETAILS_AND_PREVOUT);
     bench.run([&] {
         auto str = univalue.write();
         ankerl::nanobench::doNotOptimizeAway(str);
```

via

> build/src/bench/bench_bitcoin -filter='BlockToJsonVerbosity.*' -min-time=10000

unfortunately it reveals:
> Before:

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|        1,317,881.88 |              758.79 |    0.3% |     10.99 | `BlockToJsonVerbosity1`
|       26,534,215.46 |               37.69 |    0.3% |     10.95 | `BlockToJsonVerbosity2`
|       26,486,601.35 |               37.75 |    0.1% |     10.95 | `BlockToJsonVerbosity3`

> After:

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|        1,313,779.11 |              761.16 |    0.6% |     11.00 | `BlockToJsonVerbosity1`
|       26,128,170.14 |               38.27 |    0.2% |     11.02 | `BlockToJsonVerbosity2`
|       26,143,108.35 |               38.25 |    0.2% |     11.03 | `BlockToJsonVerbosity3`

i.e. barely any speedup for the RPC.

ismaelsadeeq replied:
> The problem is that the goal of this PR was to speed up the RPC.

I've updated the description and the title to limit the usage to just `blockToJSON`

l0rinc replied:
My concern isn't with the description (I think reserving space can be a valuable change), but that it doesn't really help the user calling this RPC, right?

ismaelsadeeq replied:
Yes end to end bench it's not significant.
I see it locally as well.

l0rinc replied:
I would be ok with it if we make the benchmarks representative - or not claim that this improved the RPC (I already have the follow-up PR locally using reserve for all other cases, I just want to be honest about it not being a measurable optimization)

ismaelsadeeq replied:
Ahh I see,
Initially I was under the impression that it did, but I was wrong hence I updated the title and description

ismaelsadeeq commented on src/bench/rpc_blockchain.cpp:
```diff
@@ -45,16 +45,33 @@ struct TestBlockAndIndex {
 } // namespace
 
-static void BlockToJsonVerbose(benchmark::Bench& bench)
+static void BlockToJson(benchmark::Bench& bench, TxVerbosity verbosity_level)
```
Done, thanks

ismaelsadeeq commented on src/core_write.cpp:
```diff
@@ -203,6 +204,7 @@ void TxToUniv(const CTransaction& tx, const uint256& block_hash, UniValue& entry
+            txinwitness.reserve(tx.vin[i].scriptWitness.stack.size());
             for (const auto& item : tx.vin[i].scriptWitness.stack) {
                 txinwitness.push_back(HexStr(item));
             }
```
good idea, pushed.
Thanks

Eunovo reviewed (approved):
Tested ACK https://github.com/bitcoin/bitcoin/pull/31179/commits/bf5c569898d0297de010102a623bf52009607ed8

Confirmed slight improvement in `BlockToJson` bench.
Before:
```
|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|          438,695.00 |            2,279.49 |    1.3% |    3,818,825.00 |    1,495,260.00 |  2.554 |     682,779.00 |    0.8% |      0.01 | `BlockToJsonVerbosity1`
|       45,518,930.00 |               21.97 |    0.3% |  354,868,701.00 |  154,939,608.00 |  2.290 |  60,169,811.00 |    1.5% |      0.51 | `BlockToJsonVerbosity2`
|       45,463,793.00 |               22.00 |    0.2% |  354,866,371.00 |  154,811,098.00 |  2.292 |  60,168,983.00 |    1.4% |      0.51 | `BlockToJsonVerbosity3`
```

After:
```
|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|          394,595.00 |            2,534.24 |    0.2% |    3,649,536.00 |    1,344,645.33 |  2.714 |     661,067.00 |    0.7% |      0.01 | `BlockToJsonVerbosity1`
|       44,967,386.00 |               22.24 |    0.1% |  353,693,003.00 |  153,066,486.00 |  2.311 |  59,949,244.00 |    1.4% |      0.51 | `BlockToJsonVerbosity2`
|       44,831,936.00 |               22.31 |    0.5% |  353,691,496.00 |  152,611,234.00 |  2.318 |  59,948,519.00 |    1.4% |      0.51 | `BlockToJsonVerbosity3`
```

maflcko reviewed (commented):
Is this worth it, when the real end-to-end speedup is less than a percent, according to https://github.com/bitcoin/bitcoin/pull/31179#discussion_r1900164051?

Edit: I see you had an end-to-end benchmark in the initial description, but it was replaced by the micro-benchmark. I guess the block in the micro-benchmark is too small to make a noticeable difference?

ismaelsadeeq commented:
> Is this worth it, when the real end-to-end speedup is less than a percent, according to https://github.com/bitcoin/bitcoin/pull/31179#discussion_r1900164051?

I think it is useful to have 1749aef52af8ea5f436e5b26dc7281fce73d1436
We can use it in places like bf5c569898d0297de010102a623bf52009607ed8 and maybe more like @l0rinc mentioned  https://github.com/bitcoin/bitcoin/pull/31179#pullrequestreview-2526455183.


> Edit: I see you had an end-to-end benchmark in the initial description, but it was replaced by the micro-benchmark. I guess the block in the micro-benchmark is too small to make a noticeable difference?

Yeah the first run of the end to end benchmark shows a noticeable difference but then subsequent runs were very slight, so the performance increase due to the reallocation is not much, hence I limit the scope of the change to just adding the reservation feature and using it in `blockToJSON`.

ismaelsadeeq commented on src/core_write.cpp:
```diff
@@ -201,9 +202,10 @@ void TxToUniv(const CTransaction& tx, const uint256& block_hash, UniValue& entry
             in.pushKV("scriptSig", std::move(o));
         }
-        if (!tx.vin[i].scriptWitness.IsNull()) {
+        if (auto witness{tx.vin[i].scriptWitness}; !witness.IsNull()) {
```
Fixed in https://github.com/bitcoin/bitcoin/compare/bf5c569898d0297de010102a623bf52009607ed8..ea62aaed3b1851605c9ff33a9d9f9edcb53828a5

Eunovo reviewed (commented):
Re-ACK https://github.com/bitcoin/bitcoin/pull/31179/commits/5d82d92aff7c11ce17ee809c060e37f73a8a687a

l0rinc commented:
While I don't think this is a measurable optimization (i.e. RPCs won't get faster because of it), it does make sense to reduce useless work and memory fragmentation.
We're reserving the right amounts - and can do other such cases (e.g. https://github.com/bitcoin/bitcoin/blob/master/src/rpc/rawtransaction.cpp#L1246) in follow-ups.

```bash
rm -rfd build && cmake -B build -DBUILD_BENCH=ON -DCMAKE_BUILD_TYPE=Release && cmake --build build -j$(nproc) && build/bin/bench_bitcoin -filter='BlockToJsonVerbosity.*'
```

> Before (bd461195f4b3a3c53366c8f197e7701836cab053):

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|          120,922.57 |            8,269.75 |    2.0% |      0.01 | `BlockToJsonVerbosity1`
|       19,956,458.00 |               50.11 |    0.3% |      0.22 | `BlockToJsonVerbosity2`
|       20,062,458.00 |               49.84 |    0.8% |      0.22 | `BlockToJsonVerbosity3`

> After (5d82d92aff7c11ce17ee809c060e37f73a8a687a):

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|          103,754.20 |            9,638.16 |    1.3% |      0.01 | `BlockToJsonVerbosity1`
|       19,978,000.00 |               50.06 |    0.6% |      0.22 | `BlockToJsonVerbosity2`
|       20,106,791.00 |               49.73 |    0.5% |      0.22 | `BlockToJsonVerbosity3`

ACK 5d82d92aff7c11ce17ee809c060e37f73a8a687a
